applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_base_element.cpp

```diff
@@ -0,0 +1,603 @@
+    {
+        // Save properties and its pointer
+        Properties& r_global_property = mpPrimalElement->GetProperties();
+        Properties::Pointer p_global_properties = mpPrimalElement->pGetProperties();
+
+        // Create new property and assign it to the element
+        Properties::Pointer p_local_property(new Properties(r_global_property));
+        mpPrimalElement->SetProperties(p_local_property);
+
+        // Disturb the design variable
+        const double current_property_value = mpPrimalElement->GetProperties()[rDesignVariable];
+        p_local_property->SetValue(rDesignVariable, (current_property_value + delta));
+
+        this->AfterPerturbation(rDesignVariable, rCurrentProcessInfo);
+
+        // Compute RHS after disturbance
+        mpPrimalElement->CalculateRightHandSide(RHS_dist, copy_process_info);
+
+        // Compute derivative of RHS w.r.t. design variable with finite differences
+        noalias(RHS_dist) -= RHS_undist;
+        RHS_dist /= delta;
+        for(unsigned int i = 0; i < RHS_dist.size(); i++)
+            rOutput(0, i) = RHS_dist[i];
+
+        // Give element original properties back
+        mpPrimalElement->SetProperties(p_global_properties);
+
+        this->AfterPerturbation(rDesignVariable, rCurrentProcessInfo);
+
+        //call one last time to make sure everything is as it was before TODO improve this..
+        mpPrimalElement->CalculateRightHandSide(RHS_dist, copy_process_info);
+    }
+    else
+        rOutput.clear();
+
+    KRATOS_CATCH("")
+
+}
+
+void AdjointFiniteDifferencingBaseElement::CalculateSensitivityMatrix(const Variable<array_1d<double,3>>& rDesignVariable, Matrix& rOutput,
+                                            const ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY;
+
+    // define working variables
+    Vector RHS_undist;
+    Vector RHS_dist;
+    ProcessInfo copy_process_info = rCurrentProcessInfo;
+
+    // Get disturbance measure
+    double delta= this->GetValue(DISTURBANCE_MEASURE);
+    double correction_factor = this->GetDisturbanceMeasureCorrectionFactor(rDesignVariable);
+    delta *= correction_factor;
+
+    if(rDesignVariable == SHAPE)
+    {
+        const int number_of_nodes = mpPrimalElement->GetGeometry().PointsNumber();
+        const unsigned int dimension = rCurrentProcessInfo.GetValue(DOMAIN_SIZE);
+        const int local_size = number_of_nodes * dimension * 2;
+
+        rOutput.resize(dimension * number_of_nodes, local_size);
+
+        // compute RHS before disturbing
+        mpPrimalElement->CalculateRightHandSide(RHS_undist, copy_process_info);
+
+        int index = 0;
+        //TODO: look that this works also for parallel computing
+        for(auto& node_i : mpPrimalElement->GetGeometry())
+        {
+            for(std::size_t coord_dir_i = 0; coord_dir_i < dimension; coord_dir_i++)
+            {
+                // disturb the design variable
+                node_i.GetInitialPosition()[coord_dir_i] += delta;
```

Comment:
We have to be very sure that we get the correct sensitivity by perturbing initial position only. For some elements doing this might give you incomplete partial sensitivity that appears correct qualitatively. I had this mistake in my code with the TL element.

Reply:
Right, this we have to investigate, i think by perturbing reference and actual, as you suggest, we are on the safe side.

Reply:
@Armin TODO check if this is replaced everywhere